### ISSUE TITLE
EPIC: Data Viewer refactor

### ISSUE THREAD

**Issue by mikebessuille** [rstudio/rstudio#7600]
Refactor the Data Viewer;  there are many bugs related to the data viewer, so we may need to refactor the component in order to clean up many issues at once.

**Comment by jmcphers** [comment]
The biggest problem with the data viewer is that it is not fully virtualized; in particular, it virtualizes rows but not columns. Visualizing wide (many columns) data sets used to bring RStudio to its knees; we eventually implemented a paging system that let you select a subset of columns, but this introduced a lot of bugs around displayed offsets, filtering (https://github.com/rstudio/rstudio/issues/6147, https://github.com/rstudio/rstudio/issues/7534), etc. 

A secondary problem is that it doesn't have a selection model, so many frequently useful features such as selecting and copying sections of data can't be implemented. 

Finally, the whole thing is glued together somewhat haphazardly from a large amount of ad-hoc JavaScript (https://github.com/rstudio/rstudio/blob/master/src/cpp/session/resources/grid/dtviewer.js) plus a third-party library (DataTables) plus an official "infinite scrolling" extension that isn't very well maintained or supported (see e.g., https://github.com/DataTables/Scroller/issues/47). 

This is why at time of writing there are [40+ bugs open against the data viewer](https://github.com/rstudio/rstudio/issues?q=is%3Aissue+is%3Aopen+label%3A%22data+viewer%22); the codebase is difficult to work on and suffers from some structural deficiencies. 

We might consider trying to find a modern alternative to DataTables that is performant with arbitrarily sized data. There are many JavaScript table visualization libraries nowadays, but most of them require all of the data to be provided up front, which is a complete nonstarter for us as we often visualize data from frames or matrices several gigabytes or more in size.

**Comment by francisbarton** [comment]
This does seem like an epic piece. My feature request for a brand new Data Viewer would be that it would have its own pane/tab, or be rendered in the Viewer tab. It has always seemed odd and inconvenient, to me, that View() opens up a tab in the Source pane.

**Comment by jmcphers** [comment]
The best existing I've found for this is the Phosphor DataGrid. https://github.com/phosphorjs/phosphor/blob/master/packages/datagrid/src/datagrid.ts

**Comment by kevinushey** [comment]
The other thing we could consider... would implementing and maintaining our own JS (or TS) table viewer be worthwhile? My worry is that any existing off-the-shelf solution might provide us with 90% of what we need, but trying to get the last 10% done will put us back into the same place we currently are, with a mixture of extensions that we've glued together that are likely to bitrot over time. Now that the team is growing, we might have the capacity to build and maintain our own solution, tailored to our own requirements.

From what I can see, there aren't any really actively maintained JS table viewers; even phosphor's DataGrid saw its last update 2 years ago:

https://www.npmjs.com/package/@phosphor/datagrid
https://github.com/phosphorjs/phosphor/blob/master/packages/datagrid/src/datagrid.ts

It's possible that it's just "done" and good enough, but that feels a bit unlikely.

**Comment by jmcphers** [comment]
Yes, I wasn't suggesting that we actually create a dependency on it, just learn from it since they clearly figured out how to actually make arbitrarily large data grids performant in the browser. Forking it might be an option, too.

**Comment by gtritchie** [comment]
Data viewer has been rewritten per: https://github.com/rstudio/rstudio/pull/17382
